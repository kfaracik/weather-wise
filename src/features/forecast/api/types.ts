export type LocationCord = {
  lat: number;
  lon: number;
};

export type CityName = {
  cityName: string;
};

export type Location = LocationCord & {
  country: string;
  name: string;
  state?: string;
};

export type WeatherCondition = {
  id: number;
  main: string;
  description: string;
  icon: string;
};

export type MainWeatherData = {
  temp: number;
  feels_like?: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
  sea_level?: number;
  grnd_level?: number;
  temp_kf?: number;
};

export type Wind = {
  speed: number;
  deg?: number;
  gust?: number;
};

export type Clouds = {
  all: number;
};

export type SysInfo = {
  country: string;
  sunrise?: number;
  sunset?: number;
  pod?: string;
};

export type BaseWeatherData = {
  dt: number;
  main: MainWeatherData;
  weather: WeatherCondition[];
  clouds: Clouds;
  wind: Wind;
  visibility: number;
  pop?: number;
  rain?: {
    '3h': number;
  };
  snow?: {
    '3h': number;
  };
  sys: SysInfo;
  dt_txt?: string;
};

export type CurrentWeatherResponse = {
  coord: LocationCord;
  weather: WeatherCondition[];
  main: MainWeatherData;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  sys: SysInfo;
  timezone: number;
  id: number;
  name: string;
  cod: number;
};

export type WeatherForecastResponse = {
  cod: string;
  message: number;
  cnt: number;
  list: BaseWeatherData[];
  city: {
    id: number;
    name: string;
    coord: LocationCord;
    country: string;
    population: number;
    timezone: number;
    sunrise: number;
    sunset: number;
  };
};
