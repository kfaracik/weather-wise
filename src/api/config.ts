export const apiKey = '9217e79516c91537d3ccd7051262951c';
export const baseUrl = 'https://api.openweathermap.org';
