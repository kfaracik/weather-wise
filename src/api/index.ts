export {apiKey, baseUrl} from './config';
