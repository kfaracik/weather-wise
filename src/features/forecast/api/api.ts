import axios from 'axios';

export const apiCall = async (endpoint: string) => {
  const options = {
    method: 'GET',
    url: endpoint,
  };

  try {
    const response = await axios.request(options);
    return response.data;
  } catch (error) {
    console.error({error});
    return null;
  }
};
