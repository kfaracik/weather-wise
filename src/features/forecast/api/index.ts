export {
  fetchLocation,
  fetchWeatherForecast,
  type CityName,
  type Location,
  type WeatherResponse,
} from './forecastService';
