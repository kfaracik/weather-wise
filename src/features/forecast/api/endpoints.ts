import {apiKey, baseUrl} from '../../../api';

export type Forecast = {
  lat: number;
  lon: number;
};

export type GetLocation = {
  cityName: string;
  limit?: number;
};

export const forecastEndpoint = ({lat, lon}: Forecast) =>
  `https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${lon}&appid=${apiKey}`;

export const getLocationEndpoint = ({cityName, limit = 5}: GetLocation) =>
  `${baseUrl}/geo/1.0/direct?q=${cityName}&limit=${limit}&appid=${apiKey}`;
