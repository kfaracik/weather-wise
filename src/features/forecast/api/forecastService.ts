import {apiCall} from './api';
import {
  currentWeatherEndpoint,
  getLocationEndpoint,
  forecastEndpoint,
} from './endpoints';
import type { CityName, CurrentWeatherResponse, Location, LocationCord, WeatherForecastResponse } from './types';

export const fetchLocation = async (
  params: CityName,
): Promise<Location[]> => {
  const response = await apiCall(getLocationEndpoint(params));
  return response as Location[];
};

export const fetchCurrentWeather = async (
  params: LocationCord,
): Promise<CurrentWeatherResponse> => {
  const response = await apiCall(currentWeatherEndpoint(params));
  return response as CurrentWeatherResponse;
};

export const fetchWeatherForecast = async (
  params: LocationCord,
): Promise<WeatherForecastResponse> => {
  const response = await apiCall(forecastEndpoint(params));
  return response as WeatherForecastResponse;
};
