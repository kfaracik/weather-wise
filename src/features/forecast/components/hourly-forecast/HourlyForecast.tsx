import React from 'react';
import { StyleSheet, View } from 'react-native';
import { DataTable, Text } from 'react-native-paper';
import { BaseWeatherData } from '../../api/types';

type HourlyForecastProps = {
  hourlyData: BaseWeatherData[];
};

export const HourlyForecast = ({ hourlyData }: HourlyForecastProps) => {
  return (
    <View style={styles.container}>
      <Text style={styles.title}>Hourly Forecast</Text>
      <DataTable>
        <DataTable.Header>
          <DataTable.Title>Time</DataTable.Title>
          <DataTable.Title>Temp</DataTable.Title>
          <DataTable.Title>Description</DataTable.Title>
        </DataTable.Header>
        {hourlyData.map((item) => (
          <DataTable.Row key={item.dt}>
            <DataTable.Cell>
              {new Date(item.dt * 1000).toLocaleTimeString([], {
                hour: '2-digit',
                minute: '2-digit',
              })}
            </DataTable.Cell>
            <DataTable.Cell>{item.main.temp}Â°C</DataTable.Cell>
            <DataTable.Cell>{item.weather[0].description}</DataTable.Cell>
          </DataTable.Row>
        ))}
      </DataTable>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    marginTop: 16,
    padding: 16,
    borderRadius: 8,
    backgroundColor: 'rgba(255, 255, 255, 0.5)',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 8,
  },
});
