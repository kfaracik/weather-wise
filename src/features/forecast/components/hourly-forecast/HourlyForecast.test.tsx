import React from 'react';
import {render} from '@testing-library/react-native';
import {HourlyForecast} from './HourlyForecast';
import {BaseWeatherData} from '../../api/types';

const mockHourlyData: BaseWeatherData[] = [
  {
    dt: 1628167200,
    main: {
      temp: 25.6,
      feels_like: 26.0,
      temp_min: 25.0,
      temp_max: 26.5,
      pressure: 1013,
      humidity: 78,
      sea_level: 1013,
      grnd_level: 1000,
      temp_kf: undefined,
    },
    weather: [
      {
        id: 800,
        main: 'Clear',
        description: 'clear sky',
        icon: '01d',
      },
    ],
    clouds: {
      all: 10,
    },
    wind: {
      speed: 5.1,
      deg: 250,
      gust: 7.0,
    },
    visibility: 10000,
    sys: {
      pod: 'd',
      country: 'PL',
    },
  },
];

describe('LocationSearchScreen', () => {
  it('should generate snapshot', () => {
    const {toJSON} = render(<HourlyForecast hourlyData={mockHourlyData} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
