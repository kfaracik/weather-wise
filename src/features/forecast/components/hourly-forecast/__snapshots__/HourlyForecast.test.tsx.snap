// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSearchScreen should generate snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(255, 255, 255, 0.5)",
      "borderRadius": 8,
      "marginTop": 16,
      "padding": 16,
    }
  }
>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "fontFamily": "System",
          "fontWeight": "400",
          "letterSpacing": 0,
        },
        {
          "writingDirection": "ltr",
        },
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "marginBottom": 8,
        },
      ]
    }
  >
    Hourly Forecast
  </Text>
  <View
    style={
      [
        {
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "paddingHorizontal": 16,
          },
          {
            "borderBottomColor": "rgba(231, 224, 236, 1)",
          },
          undefined,
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignContent": "center",
              "flex": 1,
              "flexDirection": "row",
              "paddingVertical": 12,
            },
            undefined,
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "alignItems": "center",
                  "fontSize": 12,
                  "fontWeight": "500",
                  "lineHeight": 24,
                },
                {
                  "maxHeight": 24,
                },
                {},
                {
                  "color": "rgba(28, 27, 31, 0.6)",
                },
                undefined,
              ],
            ]
          }
        >
          Time
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignContent": "center",
              "flex": 1,
              "flexDirection": "row",
              "paddingVertical": 12,
            },
            undefined,
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "alignItems": "center",
                  "fontSize": 12,
                  "fontWeight": "500",
                  "lineHeight": 24,
                },
                {
                  "maxHeight": 24,
                },
                {},
                {
                  "color": "rgba(28, 27, 31, 0.6)",
                },
                undefined,
              ],
            ]
          }
        >
          Temp
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignContent": "center",
              "flex": 1,
              "flexDirection": "row",
              "paddingVertical": 12,
            },
            undefined,
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "alignItems": "center",
                  "fontSize": 12,
                  "fontWeight": "500",
                  "lineHeight": 24,
                },
                {
                  "maxHeight": 24,
                },
                {},
                {
                  "color": "rgba(28, 27, 31, 0.6)",
                },
                undefined,
              ],
            ]
          }
        >
          Description
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          false,
          [
            {
              "borderBottomWidth": 0.5,
              "borderStyle": "solid",
              "minHeight": 48,
              "paddingHorizontal": 16,
            },
            {
              "borderBottomColor": "rgba(231, 224, 236, 1)",
            },
            undefined,
          ],
        ]
      }
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              false,
              [
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                undefined,
              ]
            }
            testID="undefined-text-container"
          >
            14:40
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              false,
              [
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                undefined,
              ]
            }
            testID="undefined-text-container"
          >
            25.6
            °C
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              false,
              [
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                undefined,
              ]
            }
            testID="undefined-text-container"
          >
            clear sky
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
